### PR TITLE
Implemented logfwdr binding

### PR DESCRIPTION
**Pull request by alankemp**
This implements a new logfwdr binding which will only be usable by accounts with a special permission.

### CODE REVIEW

File: .changeset/shaggy-poems-talk.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": minor
```
Comment (threepointone) [inline .changeset/shaggy-poems-talk.md:2]: we mark new features as "patch"

File: packages/wrangler/src/__tests__/configuration.test.ts [hunk 2]
```diff
@@ -40,6 +40,10 @@ describe("normalizeAndValidateConfig()", () => {
 			tsconfig: undefined,
 			kv_namespaces: [],
 			legacy_env: true,
+			logfwdr: {
+				bindings: [],
+				schema_filename: "",
```
Comment (threepointone) [inline packages/wrangler/src/__tests__/configuration.test.ts:45]: let's just call it "schema"

File: packages/wrangler/src/config/environment.ts [hunk 3]
```diff
@@ -338,6 +338,29 @@ interface EnvironmentNonInheritable {
...
 
+	/**
+	 * Specify a compiled capnp schema to use
+	 * Then add a binding per field in the top level message that you will send to logfwdr
+	 *
+	 * NOTE: This field is not automatically inherited from the top level environment,
+	 * and so must be specified in every named environment.
+	 *
+	 * @default `{schema:undefined,bindings:[]}`
+	 * @nonInheritable
```
Comment (threepointone) [inline packages/wrangler/src/config/environment.ts:349]: why shouldn't this be inherited?

File: packages/wrangler/src/config/validation.ts [hunk 4]
```diff
@@ -1410,6 +1423,46 @@ const validateDurableObjectBinding: ValidatorFn = (
...
+		return false;
+	}
+
+	let isValid = true;
+	if (!isRequiredProperty(value, "name", "string")) {
+		diagnostics.errors.push(`binding should have a string "name" field.`);
+		isValid = false;
+	}
+
+	// TODO: not sure how to write this
```
Comment (threepointone) [inline packages/wrangler/src/config/validation.ts:1440]: I'll have a look at this later

**Review by threepointone — CHANGES_REQUESTED** [review]
This is looking good! I'm going to just block it for today so I can have a proper look before landing it, but it's looking straight forward. Dropped some initial notes.

File: packages/wrangler/src/__tests__/publish.test.ts [hunk 5]
```diff
@@ -4223,6 +4223,21 @@ addEventListener('fetch', event => {});`
 					DATA_BLOB_ONE: "./some-data-blob.bin",
 					DATA_BLOB_TWO: "./more-data-blob.bin",
 				},
+				logfwdr: {
+					schema_filename: "./message.capnp.compiled",
+					bindings: [
+						{
+							name: "httplogs",
+							union_member_name: "httplogs",
+							union_member_tag: 1,
```
Comment (threepointone) [inline packages/wrangler/src/__tests__/publish.test.ts:4232]: lemme ping you tomorrow to understand what this tag/name does

File: .gitignore [hunk 6]
```diff
@@ -176,6 +176,7 @@ dist
 wrangler-dist/
 miniflare-dist
 packages/wrangler/wrangler.toml
+packages/wrangler/message.capnp.compiled
```
Comment (threepointone) [inline .gitignore:179]: odd. this shouldn't even be necessary, because we run the publish tests in a temp dir.

Comment (alankemp) [reply .gitignore:179]: I might be doing something wrong, but I didn't see any other tests explicitly changing the directory?

Comment (JacobMGEvans) [reply .gitignore:179]: `runInTemp` has everything in the test run in a directory that is removed. If you need examples of a test that changes directories during a test closure, anything using `setWorkingDirectory("some-name")` is doing a `process.chdir()`

File: packages/wrangler/src/__tests__/configuration.test.ts [hunk 7]
```diff
@@ -40,6 +40,10 @@ describe("normalizeAndValidateConfig()", () => {
 			tsconfig: undefined,
 			kv_namespaces: [],
 			legacy_env: true,
+			logfwdr: {
+				bindings: [],
+				schema: "",
```
Comment (threepointone) [inline packages/wrangler/src/__tests__/configuration.test.ts:45]: should default here be undefined?
```suggestion
				schema: undefined,
```

File: packages/wrangler/src/config/environment.ts [hunk 8]
```diff
@@ -338,6 +338,27 @@ interface EnvironmentNonInheritable {
...
+	 */
+	logfwdr: {
+		/** capnp schema filename */
+		schema: string;
```
Comment (threepointone) [inline packages/wrangler/src/config/environment.ts:353]:
```suggestion
		schema: string | undefined;
```
as specified a couple of lines above here

File: packages/wrangler/src/config/validation.ts [hunk 9]
```diff
@@ -1037,6 +1037,19 @@ function normalizeAndValidateEnvironment(
...
+		logfwdr: notInheritable(
+			diagnostics,
+			topLevelEnv,
+			rawConfig,
+			rawEnv,
+			envName,
+			"logfwdr",
+			validateBindingsProperty(envName, validateCflogfwdrBinding),
+			{
+				schema: "",
```
Comment (threepointone) [inline packages/wrangler/src/config/validation.ts:1049]:
```suggestion
				schema: undefined,
```

File: packages/wrangler/src/index.tsx [hunk 10]
```diff
@@ -938,6 +938,7 @@ function createCLIParser(argv: string[]) {
 											text_blobs: {},
 											data_blobs: {},
 											worker_namespaces: [],
+											logfwdr: { schema: "", bindings: [] },
```
Comment (threepointone) [inline packages/wrangler/src/index.tsx:941]:
```suggestion
											logfwdr: { schema: undefined, bindings: [] },
```

**Review by threepointone — COMMENTED** [review]
Dropped some notes. Trying this out locally to see why the gitgnore line is required

File: .gitignore (hunk 6 above)
Comment (threepointone) [reply .gitignore:179]: I tried this out locally, and it doesn't output directly into this directory, so isn't needed. I'm going to remove it myself.

**Review by threepointone — APPROVED** [review]
Looks good. I'll land this once CI passes.

**Pull request merged**